python/lsst/daf/butler/registry/databases/postgresql.py

```diff
@@ -55,26 +56,36 @@ class PostgresqlDatabase(Database):
     PostgreSQL server is installed and can be run locally in userspace.
     """
 
-    def __init__(self, *, origin: int, uri: str, namespace: Optional[str] = None, writeable: bool = True):
-        engine = sqlalchemy.engine.create_engine(uri, pool_size=1)
-        dbapi = engine.raw_connection()
+    def __init__(self, *, cs: ConnectionStruct, origin: int, namespace: Optional[str] = None,
+                 writeable: bool = True):
+        super().__init__(origin=origin, cs=cs, namespace=namespace)
+        dbapi = cs.engine.raw_connection()
         try:
             dsn = dbapi.get_dsn_parameters()
         except (AttributeError, KeyError) as err:
             raise RuntimeError(f"Only the psycopg2 driver for PostgreSQL is supported.") from err
         if namespace is None:
-            namespace = dsn.get("schema")
-            if namespace is None:
-                raise RuntimeError("Namespace (schema) must be provided as an argument or included in URI.")
-        else:
-            if dsn.get("schema", namespace) != namespace:
-                raise RuntimeError("'namespace' and 'uri' arguments specify different schemas.")
-        super().__init__(origin=origin, engine=engine, namespace=namespace)
-        if not writeable:
-            dbapi.set_session(readonly=True)
-        self.dbname = dsn["dbname"]
+            namespace = dsn["schema"]
+        self.namespace = namespace
+        self.dbname = dsn.get("dbname")
         self._writeable = writeable
-        self._shrinker = NameShrinker(engine.dialect.max_identifier_length)
+        self._shrinker = NameShrinker(cs.engine.dialect.max_identifier_length)
+
+    @classmethod
+    def connect(cls, uri: str, *, writeable: bool = True) -> ConnectionStruct:
+        cs = ConnectionStruct(sqlalchemy.engine.create_engine(uri, pool_size=1))
+        if not writeable:
+            dbapi = cs.engine.raw_connection()
+            try:
+                dbapi.set_session(readonly=True)
```

Comment:
I wonder if this works 100% reliably. I understand that connections are cached in a pool, but it may also happen that pool can decide to drop and re-establish a connection under some circumstances, and new connection will not have `readonly` option set. It may be better to pass a `creator` argument to `create_engine` or use `SET TRANSACTION READ ONLY` when transaction is started.